### PR TITLE
Add functional test for DP coordinator throughput

### PR DESCRIPTION
**Pull request by tdene**
# What does this PR do ?
This PR adds a test that runs a multi-DP workload to test the throughput of the inference dynamic coordinator.

BENCHMARK INFO: The "golden" benchmark returns an average value of `80.21`, with expected noise of +/- 2%.
This PR establishes the first known value of the "golden" benchmark.

The "unit-test" benchmark returns average values as seen below, with expected noise of +/- 5%.
```
Initialization time: 4502.31 ms
Run time: 3065.60 ms
Stop time: 127.12 ms
ZMQ throughput is approximately 32.62 requests/ms
```
This PR establishes the first known value of the "unit-test" benchmark.

WHY IS THIS NECESSARY: This functional tests provides a "golden" benchmark that other PRs use to determine whether they have degraded or improved throughput. For example, PRs which quote this "golden" benchmark are: #XXXX.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: tests/functional_tests/test_cases/gpt/gpt_dynamic_inference_tp8_pp1_dp1_583m_latencytest_zmq/golden_values_dev_dgx_h100.json [hunk 1]
```diff
@@ -0,0 +1,3 @@
+{
+ "throughput": 0.8
```
Comment (mathemakitten) [inline tests/functional_tests/test_cases/gpt/gpt_dynamic_inference_tp8_pp1_dp1_583m_latencytest_zmq/golden_values_dev_dgx_h100.json]: Is this a median value taken from N real trials? What's the variation on it? We've previously had issues with small throughput values causing issues across different test clusters.

Comment (tdene) [reply tests/functional_tests/test_cases/gpt/gpt_dynamic_inference_tp8_pp1_dp1_583m_latencytest_zmq/golden_values_dev_dgx_h100.json]: It was a placeholder; real value is in now.

I have not tested variation thoroughly, but subjectively it seems pretty low. It's averaged out across 128 requests so that's probably what keeps it low.

Comment (tdene) [reply tests/functional_tests/test_cases/gpt/gpt_dynamic_inference_tp8_pp1_dp1_583m_latencytest_zmq/golden_values_dev_dgx_h100.json]: Update: the variation is too high, even when averaged out across 1024 requests. Because I am randomly generating prompts on the fly.

I think the only correct solution here is to add a dataset of deterministic prompts. I don't know how to do that today, so I will circle back and learn how to do it later.

Comment (mathemakitten) [reply tests/functional_tests/test_cases/gpt/gpt_dynamic_inference_tp8_pp1_dp1_583m_latencytest_zmq/golden_values_dev_dgx_h100.json]: OK! You might want to reuse the existing prompt dataset we use in other inference tests which is already mounted. See `tests/functional_tests/test_cases/gpt/gpt_dynamic_inference_tp1_pp1_583m_logitsmatch/model_config.yaml`.

Comment (tdene) [reply tests/functional_tests/test_cases/gpt/gpt_dynamic_inference_tp8_pp1_dp1_583m_latencytest_zmq/golden_values_dev_dgx_h100.json]: Done! Can I get a review/approve @mathemakitten

**Comment by tdene** [comment]
> Could you share an internal pipeline id that validates this test, once this is ready?

@ko3n1g `38639505`.

**Comment by tdene** [comment]
> Could you share an internal pipeline id that validates this test, once this is ready?

@ko3n1g my other message got lost in all the copy-pr-bot messages, so I'll repost here.    
`38639505` is the ID.

**Comment by kvareddy** [comment]
@mathemakitten  do you have any other concerns? Once you approve the MR, I can take a final look.

File: examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py [hunk 2]
```diff
@@ -61,27 +61,41 @@ async def main(
...
+            if args.incoming_requests_per_step is None:
+                # Only add requests that have arrived at the current time.
+                while num_requests_added < num_requests_total and requests[num_requests_added].time_arrival <= current_time:
+                    request = requests[num_requests_added]
+                    # These add-request calls will queue up the request on a zmq socket and return
+                    # instantaneously. They will return an asyncio future which can be awaited for
+                    # request completion.
+                    futures.append(client.add_request(request.prompt_text, request.sampling_params))
+                    num_requests_added += 1
+                    #tbar.update(1)
```
Comment (mathemakitten) [inline examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py]: Remove commented code?

File: examples/inference/gpt/utils.py [hunk 3]
```diff
@@ -299,9 +305,12 @@ def get_requests_from_file(
     # Load prompts.
     n_prompts = sum(1 for _ in open(args.prompt_file))
     prompts = []
+    num_tokens_to_generate = []
     with open(args.prompt_file) as f:
         for line in tqdm(f.readlines(), "read prompt file", total=n_prompts):
             prompts.append(json.loads(line)["text"])
+            if args.num_tokens_from_file:
+                num_tokens_to_generate.append(json.loads(line)["chatgpt_output_token_length"])
```
Comment (mathemakitten) [inline examples/inference/gpt/utils.py]: nit: suggest `l = json.loads(line)` and something like `if l.get("chatgpt_output_token_length"): num_tokens_to_generate.append(l["chatgpt_output_token_length"])`? Would also save us the extra CLI flag and also skip having to call `json.loads(line)` twice.

File: examples/inference/gpt/utils.py [hunk 4]
```diff
@@ -313,10 +322,20 @@ def get_requests_from_file(
...
 
+    # Update sampling params with per-prompt values.
+    sampling_params_list = [sampling_params] * len(prompts)
+    if args.num_tokens_from_file:
+        if sampling_params is None:
+            sampling_params = get_default_sampling_params(tokenizer.eod)
+        for i in range(len(prompts)):
+            sp = copy.deepcopy(sampling_params)
+            sp.num_tokens_to_generate = num_tokens_to_generate[i]
+            sampling_params_list[i] = sp
```
Comment (mathemakitten) [inline examples/inference/gpt/utils.py]: Why not do this in the file-processing loop above to avoid having to iterate through `prompts` twice?

File: examples/inference/gpt/utils.py (hunk 3 above)
Comment (tdene) [reply examples/inference/gpt/utils.py]: I wanted a separate CLI flag because the `gpt_dynamic_inference_tp1_pp1_583m_logitsmatch` also loads the same file, but ignores `chatgpt_output_token_length` altogether. So to keep pre-existing functionality I needed a flag to decide whether or not to ignore that field.

Comment (tdene) [reply examples/inference/gpt/utils.py]: Done on the `json.loads(line)` dedup!

File: examples/inference/gpt/utils.py (hunk 4 above)
Comment (tdene) [reply examples/inference/gpt/utils.py]: Done!

File: examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py (hunk 2 above)
Comment (tdene) [reply examples/inference/gpt/gpt_dynamic_inference_with_coordinator.py]: Done!
